_posts/2020-06-12-solidity-0.6-inheritance.md

```diff
@@ -0,0 +1,130 @@
+
+## Explicitely `virtual` and `override`
+Functions are no longer virtual by default. This means a call to a non-virtual function will always execute that function regardless of any other contract in the inheritance hierarchy. This reduces abiguity that exists in version 0.5 where all functions are implicity virtual, allowing to be overriden further down the inheritance structure. This is especially dangerous in big inheritance graphs where this abiguity can lead to unintended behaviour and bugs.
+
+    pragma solidity ^0.5.17;
+    contract A {
+        uint public x;
+        function setValue(uint _x) public {
+            x = _x;
+        }
+    }
+
+    contract B {
+        uint public y;
+        function setValue(uint _y) public {
+            y = _y;
+        }
+    }
+
+    contract C is A, B {
+    }
+
+For example in contract `C` above calling `setValue` is calling the most derived implementation from contract `B`. However this is non-obvious from the implementation.
+
+With 0.6 for the example above the compiler will now raise a type error: `Derived contract must override function "setValue". Two or more base classes define functions with the same name and parameter types`. This forces us to rething which implementation to use and explicitely override the conflicting base ones:
+
+    pragma solidity ^0.6.10;
+    contract A {
+        uint public x;
+        function setValue(uint _x) public virtual {
+            x = _x;
+        }
+    }
+
+    contract B {
+        uint public y;
+        function setValue(uint _y) public virtual {
+            y = _y;
+        }
+    }
+
+    contract C is A, B {
+        function setValue(uint _x) public override(A,B) {
+            A.setValue(_x);
+        }
+    }
+
+You can only override functions if they are marked `virtual`. Additionally any overriding function must be marked `override`. If that again should be overridable, it has to also be marked `virtual`.
+
+Note: Interface functions are implicitely virtual so when implementing an interface you have to explicitely override its functions in the implementation. There is an ongoing discussion on this design [here](https://github.com/ethereum/solidity/issues/8281).
+
+## Abstract contracts
+With version 0.5 contract that doesn't implement all of its functions is implicitely made abstract by the compiler:
+    
+    pragma solidity ^0.5.17;
+    contract X {
+        function setValue(uint _x) public virtual;
+    }
+
+With 0.6 this differentiation has to be explicit with the compiler generating error `contract X should be made abstract` 
+    
+    pragma solidity ^0.6.10;
+    abstract contract X {
+        function setValue(uint _x) public virtual;
+    }
+
+## Interfaces can inherit
+This functionality is new to version 0.6 and allows interface inheritance. The resulting interface is a superset of all inherited interfaces' functions which the contract has to implement.
```

Comment:
"a superset" sounds like it could also include other functions. What about something like "the combination of "?